function-memoization/src/main/java/com/tomtom/kotlin/FunctionMemoizationExtensions.kt

```diff
@@ -0,0 +1,141 @@
+            this@memoize(pair.first, pair.second)
+        }.memoize(cacheSize = cacheSize)
+
+        override fun invoke(p1: P1, p2: P2): R = cache.invoke(Pair(p1, p2))
+    }
+
+/**
+ * Creates a memoized function that will return cached result when the same inputs occurs again.
+ */
+fun <P1, P2, R> ((P1, P2) -> R).memoize(): (P1, P2) -> R =
+    object : (P1, P2) -> R {
+        private val cache = { pair: Pair<P1, P2> ->
+            this@memoize(pair.first, pair.second)
+        }.memoize()
+
+        override fun invoke(p1: P1, p2: P2): R = cache.invoke(Pair(p1, p2))
+    }
+
+
+/**
+ * Creates a LRU style cached function that will return cached result when the same inputs occurs again.
+ */
+fun <P1, P2, P3, R> ((P1, P2, P3) -> R).memoize(cacheSize: Int = 1): (P1, P2, P3) -> R =
+    object : (P1, P2, P3) -> R {
+        private val cache = { triple: Triple<P1, P2, P3> ->
+            this@memoize(triple.first, triple.second, triple.third)
+        }.memoize(cacheSize = cacheSize)
+
+        override fun invoke(p1: P1, p2: P2, p3: P3): R = cache.invoke(Triple(p1, p2, p3))
+    }
+
+/**
+ * Creates a memoized function that will return cached result when the same inputs occurs again.
+ */
+fun <P1, P2, P3, R> ((P1, P2, P3) -> R).memoize(): (P1, P2, P3) -> R =
+    object : (P1, P2, P3) -> R {
+        private val cache = { triple: Triple<P1, P2, P3> ->
+            this@memoize(triple.first, triple.second, triple.third)
+        }.memoize()
+
+        override fun invoke(p1: P1, p2: P2, p3: P3): R = cache.invoke(Triple(p1, p2, p3))
+    }
+
+/**
+ * Creates a LRU style cached function that will return cached result when the same inputs occurs again.
+ */
+fun <P1, P2, P3, P4, R> ((P1, P2, P3, P4) -> R).memoize(cacheSize: Int): (P1, P2, P3, P4) -> R =
+    object : (P1, P2, P3, P4) -> R {
+        private val cache = { quadruple: Quadruple<P1, P2, P3, P4> ->
+            this@memoize(quadruple.first, quadruple.second, quadruple.third, quadruple.fourth)
+        }.memoize(cacheSize = cacheSize)
+
+        override fun invoke(p1: P1, p2: P2, p3: P3, p4: P4): R = cache.invoke(Quadruple(p1, p2, p3, p4))
+    }
+
+/**
+ * Creates a memoized function that will return cached result when the same inputs occurs again.
+ */
+fun <P1, P2, P3, P4, R> ((P1, P2, P3, P4) -> R).memoize(): (P1, P2, P3, P4) -> R =
+    object : (P1, P2, P3, P4) -> R {
+        private val cache = { quadruple: Quadruple<P1, P2, P3, P4> ->
+            this@memoize(quadruple.first, quadruple.second, quadruple.third, quadruple.fourth)
+        }.memoize()
+
+        override fun invoke(p1: P1, p2: P2, p3: P3, p4: P4): R = cache.invoke(Quadruple(p1, p2, p3, p4))
+    }
+
+private class Function1Cache<P1, R>(
+    private val originalFunction: (P1) -> R,
+    // no cache size specified means no limit
+    private val cacheSize: Int? = null
+) : (P1) -> R {
+    private val map = LinkedHashMap<P1, R>(cacheSize ?: 16, 0.75f, true)
```

Comment:
Explain magic values of 16, 0.75, and true.

Reply:
Default value for access order is "false".  LinkedHashMap does not provide constructor where access order can be specified without initial capacity (which is 16 by default) and load factor (which is 0.75f by default) - thus magic values.

Reply:
Use named values to make their meaning explicit in this case.
Also explain the use of 16 and 0.75 (note that according to the documentation of Kotlin the default load factor is 0.0, not 0.75). Explain the true provides the LRU characteristic you need.

Reply:
I was using java constructor instead of kotlin one. I've switched do kotlin and I'm doing re-insertion manually (jvm implementation was handling this in very similar way). No more magic values.